src/tag.rs

```diff
@@ -0,0 +1,415 @@
+        self.has_wiki = has_wiki;
+        self
+    }
+
+    pub fn has_artist(mut self, has_artist: bool) -> Self {
+        self.has_artist = has_artist;
+        self
+    }
+}
+
+impl Client {
+    /// Returns a Stream over all the [`Tag`]s matching the search query.
+    ///
+    /// ```no_run
+    /// # use rs621::client::Client;
+    /// # use rs621::tag::Query;
+    /// use futures::prelude::*;
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() -> rs621::error::Result<()> {
+    /// let client = Client::new("https://e926.net", "MyProject/1.0 (by username on e621)")?;
+    ///
+    /// let query = Query::new()
+    ///     .name("fluffy")
+    ///     .limit(1);
+    ///
+    /// let tag_stream = client.tag_search(query);
+    /// futures::pin_mut!(tag_stream);
+    ///
+    /// while let Some(tag) = tag_stream.next().await {
+    ///     assert_eq!(tag?.name, "fluffy");
+    /// }
+    /// # Ok(()) }
+    /// ```
+    pub fn tag_search(
+        &self,
+        query: Query,
+    ) -> impl Stream<Item = Rs621Result<Tag>> + '_ + Send + Sync {
+        // TODO: There should be a way to use `try_unfold` here instead.
+        unfold(Some(query), move |query| self.tag_search_page(query))
+            .map(futures::stream::iter)
+            .flatten()
+    }
+
+    async fn tag_search_page(
+        &self,
+        query: Option<Query>,
+    ) -> Option<(Vec<Rs621Result<Tag>>, Option<Query>)> {
+        // `/tags.json` is weird and returns `{ tags: [] }` instead of the empty array when there
+        // are no results.
+        #[derive(Deserialize)]
+        #[serde(untagged, deny_unknown_fields)]
+        enum MaybeTags {
+            #[allow(dead_code)]
+            Empty {
+                tags: (),
+            },
+            Items(Vec<Tag>),
+        }
+
+        // `query` will be `None` if the previous `tag_search_page` errored. If that is the case,
+        // this run of `tag_search_page` will return `None` to end the stream.
+        let mut query = query?;
+
+        let tags = match self.get_json_endpoint_query("/tags.json", &query).await {
+            Err(e) => return Some((vec![Err(e)], None)),
+            Ok(MaybeTags::Empty { .. }) => return None,
+            Ok(MaybeTags::Items(i)) if i.is_empty() => return None,
+            Ok(MaybeTags::Items(i)) => i,
+        };
+
+        let (min, max) = tags.iter().map(|x| x.id).minmax().into_option().unwrap();
+
+        let next_page = match query.page {
+            Some(Cursor::Before(_)) => Cursor::Before(min),
+            Some(Cursor::After(_)) => Cursor::After(max),
+            Some(Cursor::Page(p)) => Cursor::Page(p + 1),
+            None => match query.order {
+                None | Some(Order::IdDsc | Order::Date) => Cursor::Before(min),
+                Some(Order::IdAsc) => Cursor::After(max),
+                _ => Cursor::Page(2),
+            },
+        };
+        query.page = Some(next_page);
```

Comment:
So this is the part I'm least sure about. I don't fully understand the whole `a1234` / `b1234` thing and how it relates to `search[order]`.